src/vdaf/hits.rs

```diff
@@ -0,0 +1,741 @@
+                "desired bit length ({} bits) exceeds data length ({} bytes)",
+                len,
+                data.len()
+            )));
+        }
+
+        Ok(Self { data, len })
+    }
+
+    /// Return the `bit`-th bit of the input. Bounds checking is performed by the caller.
+    fn bit(&self, bit: usize) -> usize {
+        let i = bit >> 3;
+        let j = ((i << 3) ^ bit) as u8;
+        (self.data[i] >> j) as usize & 1
+    }
+
+    /// Return the index of input `self` in the look-up table of a `ToyIdpf` instance.
+    fn index(&self) -> usize {
+        let mut index = 1 << self.len;
+        for i in 0..self.len {
+            index += self.bit(i) << i;
+        }
+        index - 1
+    }
+
+    /// Construct a new input that is a prefix of `self`. Bounds checking is performed by the
+    /// caller.
+    fn prefix(&self, len: usize) -> Self {
+        Self {
+            data: self.data,
+            len,
+        }
+    }
+}
+
+impl<'a> PartialEq<IdpfInput<'a>> for IdpfInput<'a> {
+    fn eq(&self, other: &IdpfInput) -> bool {
+        if self.len != other.len {
+            return false;
+        }
+
+        // Do a constant-time comparison, masking the last byte to fit the input length (in bits).
+        let q = self.len / 8;
+        let m = (self.len % 8) as u8;
+        let mut out = 0;
+        for (i, (a, b)) in self.data.iter().zip(other.data).enumerate() {
+            let (x, y) = match i == q {
+                true => (*a & m, *b & m),
+                false => (*a, *b),
+            };
+            out |= x ^ y;
+        }
+        out == 0
+    }
+}
+
+/// An Incremental Distributed Point Function (IDPF), as defined by [[BBCG+21]].
+///
+/// [BBCG+21]: https://eprint.iacr.org/2021/017
+//
+// NOTE(cjpatton) The real IDPF API probably needs to be stateful.
+pub trait Idpf<const KEY_LEN: usize, const OUT_LEN: usize>: Sized {
+    /// The finite field over which the IDPF is defined.
+    //
+    // NOTE(cjpatton) The IDPF of [BBCG+21] might use different fields for different levels of the
+    // prefix tree.
+    type Field: FieldElement;
+
+    /// Generate a new sequence of IDPF shares for `input`. Parameter `output` is an iterator that
+    /// is invoked to get the output value for each successive level of the prefix tree.
+    fn gen<O: IntoIterator<Item = Self::Field>>(
+        input: &IdpfInput,
+        output: O,
+    ) -> Result<[Self; KEY_LEN], VdafError>;
+
+    /// Evaluate an IDPF share on `prefix`.
+    fn eval(&self, prefix: &IdpfInput) -> Result<[Self::Field; OUT_LEN], VdafError>;
+}
+
+/// A "toy" IPDF used for demonstration purposes. The space consumed by each share is `O(2^n)`,
+/// where `n` is the length of the input. The size of each share is restricted to 1MB, so this IDPF
+/// is only suitable for very short inputs.
+//
+// NOTE(cjpatton) It would be straight-forward to generalize this construction to any `KEY_LEN` and
+// `OUT_LEN`.
+pub struct ToyIdpf<F: FieldElement> {
```

Comment:
[Put the trait bounds on the impl, not the struct](https://rust-lang.github.io/api-guidelines/future-proofing.html#data-structures-do-not-duplicate-derived-trait-bounds-c-struct-bounds)
```suggestion
pub struct ToyIdpf<F> {
```